core-bundle/src/Resources/contao/pages/PageLogout.php

```diff
@@ -56,14 +55,21 @@ public function getResponse($objPage)
 			return new RedirectResponse($strRedirect);
 		}
 
+		$pairs = array();
 		$strLogoutUrl = $container->get('security.logout_url_generator')->getLogoutUrl();
-		$uri = Http::createFromString($strLogoutUrl);
+		$request = Request::create($strLogoutUrl);
+
+		if ($request->server->has('QUERY_STRING'))
+		{
+			parse_str($request->server->get('QUERY_STRING'), $pairs);
+		}
 
 		// Add the redirect= parameter to the logout URL
-		$query = new Query($uri->getQuery());
-		$query = $query->merge('redirect=' . $strRedirect);
+		$pairs['redirect'] = $strRedirect;
+
+		$request->server->set('QUERY_STRING', http_build_query($pairs, '', '&', PHP_QUERY_RFC3986));
```

Comment:
Not sure whether modifying the request object is the best solution here?

Reply:
Maybe we could use `parse_url(…, PHP_URL_QUERY)` instead of the Request class and string replace the query afterwards?

Reply:
The "replacing the query afterwards" is the problematic part here. The Symfony `Request` class is able to create a URI from the `parse_url()` parts, that's why I have used it here. Do you know a better way?

Reply:
Something like this?
```php
$url = preg_replace('(\?'.preg_quote($oldQuery).'(?=#|$))', '?'.$newQuery, $url);
```

Reply:
I don't think that this is "better" than using the request object. 🙈 @contao/developers WDYT?

Reply:
`Request::create()` accepts `$_SERVER` like arguments, so you could pass it in there.

Reply:
I have already tried this with `$request->duplicate(…, ['QUERY_STRING' => '…'])`, however, this will replace the `$server` variable instead of merge my array.

Reply:
> I don't think that this is "better" than using the request object.

You’re right, it is probably not better ☺️

The current solution is fine then I think. 👍

Reply:
I have made a small adjustment in 8da7cde8cadf4a44a2959adb164060027bfb605c, so we no longer have to overwrite the `QUERY_STRING` variable.